contracts/paymaster/TokenPaymaster.sol

```diff
@@ -0,0 +1,147 @@
+pragma solidity ^0.5.16;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/token/ERC20/IERC20.sol";
+import "./IUniswap.sol";
+import "../BasePaymaster.sol";
+
+/**
+ * A Token-based paymaster.
+ * - each request is paid for by the caller.
+ * - acceptRelayedCall - verify the caller can pay for the request in tokens.
+ * - preRelayedCall - pre-pay the maximum possible price for the tx
+ * - postRelayedCall - refund the caller for the unused gas
+ */
+contract TokenPaymaster is BasePaymaster {
+
+    //TODO: just expose them for debugging..
+    event Transfer(address indexed from, address indexed to, uint256 value);
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+
+    IUniswap public uniswap;
+    IERC20 public token;
+
+    constructor(IUniswap _uniswap) public {
+        uniswap = _uniswap;
+        token = IERC20(uniswap.tokenAddress());
+        token.approve(address(uniswap), uint(-1));
+    }
+
+    //return the payer of this request.
+    // for account-based target, this is the target account.
+    function getPayer(GSNTypes.RelayRequest calldata relayRequest) external pure returns (address) {
+        return relayRequest.target;
+    }
```

Comment:
because while in most cases the "payer" is the target contract, there are cases (e.g. mixer..) where the payer is the sender. if I hard-code it, then that TokenPaymaster can't be used as a base-class  for such scenarios.